src/dex/bebop/bebop.ts

```diff
@@ -0,0 +1,744 @@
+import { BebopConfig } from './config';
+import { Interface } from 'ethers/lib/utils';
+import { RateFetcher } from './rate-fetcher';
+import {
+  BEBOP_API_URL,
+  BEBOP_AUTH_NAME,
+  BEBOP_GAS_COST,
+  BEBOP_INIT_TIMEOUT_MS,
+  BEBOP_PRICES_CACHE_TTL,
+  BEBOP_QUOTE_TIMEOUT_MS,
+  BEBOP_TOKENS_CACHE_TTL,
+  BEBOP_TOKENS_POLLING_INTERVAL_MS,
+  BEBOP_WS_API_URL,
+} from './constants';
+import BigNumber from 'bignumber.js';
+import { getBigNumberPow } from '../../bignumber-constants';
+import { utils } from 'ethers';
+import qs from 'qs';
+
+export class Bebop extends SimpleExchange implements IDex<BebopData> {
+  readonly hasConstantPriceLargeAmounts = false;
+  readonly needWrapNative = false;
+
+  readonly isFeeOnTransferSupported = false;
+  readonly isStatePollingDex = true;
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(BebopConfig);
+
+  private rateFetcher: RateFetcher;
+  private tokensMap: TokenDataMap = {};
+
+  private pricesCacheKey: string;
+  private tokensCacheKey: string;
+  private tokensAddrCacheKey: string;
+
+  private bebopAuthToken: string;
+
+  logger: Logger;
+
+  constructor(
+    readonly network: Network,
+    readonly dexKey: string,
+    readonly dexHelper: IDexHelper,
+    readonly settlementAddress: string = BebopConfig['Bebop'][network]
+      .settlementAddress,
+    protected settlementInterface = new Interface(settlementABI),
+  ) {
+    super(dexHelper, dexKey);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.tokensCacheKey = `tokens`;
+    this.pricesCacheKey = `prices`;
+    this.tokensAddrCacheKey = `tokens_addr`;
+    const token = this.dexHelper.config.data.bebopAuthToken;
+    if (!token) {
+      throw new Error('Bebop auth token is not set');
+    }
+    this.bebopAuthToken = token;
+
+    this.rateFetcher = new RateFetcher(
+      this.dexHelper,
+      this.dexKey,
+      this.network,
+      this.logger,
+      {
+        rateConfig: {
+          tokensIntervalMs: BEBOP_TOKENS_POLLING_INTERVAL_MS,
+          pricesCacheKey: this.pricesCacheKey,
+          pricesCacheTTLSecs: BEBOP_PRICES_CACHE_TTL,
+          tokensCacheKey: this.tokensCacheKey,
+          tokensAddrCacheKey: this.tokensAddrCacheKey,
+          tokensCacheTTLSecs: BEBOP_TOKENS_CACHE_TTL,
+          tokensReqParams: {
+            url:
+              BEBOP_API_URL +
+              `/pmm/${BebopConfig['Bebop'][network].chainName}/v3/token-info`,
+          },
+          pricesReqParams: {
+            url:
+              BEBOP_WS_API_URL +
+              `/pmm/${BebopConfig['Bebop'][network].chainName}/v3/pricing`,
+            headers: {
+              name: BEBOP_AUTH_NAME,
+              authorization: this.bebopAuthToken,
+            },
+          },
+        },
+      },
+    );
+  }
+
+  async initializePricing(blockNumber: number) {
+    if (!this.dexHelper.config.isSlave) {
+      this.rateFetcher.start();
+      await sleep(BEBOP_INIT_TIMEOUT_MS);
```

Comment:
@JackieJoo Could you please explain why we need this timeout here?

Reply:
it a websocket connection, so we don't quite know when there's initial state polling finished. I don't think we need to add some complex logic for event emitting on initial state. A buffer of around 3 secs is enough. Just to make sure that the data is in cache. Overall it's mostly for testing, because in reality, we'll just see pricing after a few seconds after initial deployment.